connection.go

```diff
@@ -1666,3 +1719,43 @@ func (conn *Connection) ServerProtocolInfo() ProtocolInfo {
 func (conn *Connection) ClientProtocolInfo() ProtocolInfo {
 	return clientProtocolInfo.Clone()
 }
+
+func shutdownEventCallback(event WatchEvent) {
+	// Receives "true" on server shutdown.
+	// See https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/graceful_shutdown/
+	// step 2.
+	val, ok := event.Value.(bool)
+	if ok && val {
+		// defer cause otherwise we'll block ourselves on ack wait.
+		defer event.Conn.processShutdown()
+	}
+}
+
+func (conn *Connection) processShutdown() {
+	// Forbid starting new tasks.
+	atomic.StoreUint32(&(conn.shutdownInProgress), shutdownInProgress)
+
+	// After finish, allow starting new tasks, they will fail with
+	// "not connected" instead.
+	defer atomic.StoreUint32(&(conn.shutdownInProgress), shutdownNotInProgress)
+
+	// Wait for tasks to finish.
+	conn.shutdownWg.Wait()
+
+	// Do not unregister task explicitly since connection teardown
+	// has the same effect.
+
+	if !conn.ClosedNow() {
+		err := ClientError{
+			ErrConnectionClosed,
+			"connection closed after server shutdown",
+		}
+
+		// Start to reconnect based on common rules, same as in net.box.
+		// Reconnect also closes the connection: server waits until all
+		// subscribed connections are terminated.
+		// See https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/graceful_shutdown/
+		// step 3.
+		defer conn.reconnect(err, conn.c)
```

Comment:
```suggestion
		Why is there a `defer`?
```

Reply:
Event callback shouldn't wait for reconnect, especially since reconnect trigger re-registraction on success reconnect.

Reply:
How does defer help to achive the goal?

https://go.dev/play/p/B9iPEc3NV1g

Reply:
Thank you. I've never thought about it. Is starting reconnect is separate goroutine is better in that case?

Reply:
I think yes, it is better to do in a separate goroutine.

Reply:
In the end we decided to handle `processShutdown()` in a separate goroutine.